core/src/main/scala/kafka/server/metadata/MetadataPartitions.scala

```diff
@@ -0,0 +1,282 @@
+        }
+      }
+    }
+  }
+
+  def addUuidMapping(name: String, id: Uuid): Unit = {
+    newIdMap.put(id, name)
+  }
+
+  def removeUuidMapping(id: Uuid): Unit = {
+    newIdMap.remove(id)
+  }
+
+  def get(topicName: String, partitionId: Int): Option[MetadataPartition] = {
+    Option(newNameMap.get(topicName)).flatMap(m => Option(m.get(partitionId)))
+  }
+
+  def set(partition: MetadataPartition): Unit = {
+    val prevPartitionMap = newNameMap.get(partition.topicName)
+    val newPartitionMap = if (prevPartitionMap == null) {
+      val m = new util.HashMap[Int, MetadataPartition](1)
+      changed.put(m, true)
+      m
+    } else if (changed.containsKey(prevPartitionMap)) {
+      prevPartitionMap
+    } else {
+      val m = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() + 1)
+      m.putAll(prevPartitionMap)
+      changed.put(m, true)
+      m
+    }
+    val prevPartition = newPartitionMap.put(partition.partitionIndex, partition)
+    if (partition.isReplicaFor(brokerId)) {
+      _localChanged.add(partition)
+    } else if (prevPartition != null && prevPartition.isReplicaFor(brokerId)) {
+      _localRemoved.add(prevPartition)
+    }
+    newNameMap.put(partition.topicName, newPartitionMap)
+  }
+
+  def remove(topicName: String, partitionId: Int): Unit = {
+    val prevPartitionMap = newNameMap.get(topicName)
+    if (prevPartitionMap != null) {
+      if (changed.containsKey(prevPartitionMap)) {
+        val prevPartition = prevPartitionMap.remove(partitionId)
+        if (prevPartition.isReplicaFor(brokerId)) {
+          _localRemoved.add(prevPartition)
+        }
+      } else {
+        Option(prevPartitionMap.get(partitionId)).foreach { prevPartition =>
+          if (prevPartition.isReplicaFor(brokerId)) {
+            _localRemoved.add(prevPartition)
+          }
+          val newPartitionMap = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() - 1)
+          prevPartitionMap.forEach(new BiConsumer[Int, MetadataPartition]() {
+            override def accept(key: Int, value: MetadataPartition): Unit =
+              if (!key.equals(partitionId)) {
+                newPartitionMap.put(key, value)
+              }
+          })
+          changed.put(newPartitionMap, true)
+          newNameMap.put(topicName, newPartitionMap)
+        }
+      }
+    }
+  }
+
+  def build(): MetadataPartitions = {
+    val result = new MetadataPartitions(newNameMap, newIdMap)
+    newNameMap = Collections.unmodifiableMap(newNameMap)
+    newIdMap = Collections.unmodifiableMap(newIdMap)
+    result
+  }
+
+  def localChanged(): Set[MetadataPartition] = _localChanged.asScala.toSet
+
+  def localRemoved(): Set[MetadataPartition] = _localRemoved.asScala.toSet
+}
+
+case class MetadataPartitions(private val nameMap: util.Map[String, util.Map[Int, MetadataPartition]],
+                              private val idMap: util.Map[Uuid, String]) {
```

Comment:
Intuitively, this seems a little backwards. I would have expected to have the partitions keyed by topicId with a separate map from topic name to topic id. If we were going to support renaming for example, then we would just update the name to id map.

I suspect we are ultimately going to have to change this because it gives us no good way to lookup the UUID of a given topic name. We will need to be able to do this to handle `Metadata` queries (see https://github.com/apache/kafka/commits/trunk/core/src/main/scala/kafka/server/MetadataCache.scala).